src/Evolution/Systems/GeneralizedHarmonic/Characteristics.cpp

```diff
@@ -0,0 +1,191 @@
+    const tnsr::aa<DataVector, Dim, Frame>& u_minus,
+    const tnsr::i<DataVector, Dim, Frame>& unit_normal_one_form) noexcept {
+  auto evolved_fields =
+      make_with_value<typename Tags::EvolvedFieldsFromCharacteristicFields<Dim,
+          Frame>::type>(get(gamma_2), 0.);
+
+  // Invert Eq.(32) of Lindblom+ (2005) for Psi
+  get<Tags::Psi<Dim, Frame>>(evolved_fields) = u_psi;
+
+  for (size_t mu = 0; mu < Dim + 1; ++mu) {
+    for (size_t nu = 0; nu < mu + 1; ++nu) {
+      // Invert Eq.(32) - (34) of Lindblom+ (2005) for Pi and Phi
+      get<Tags::Pi<Dim, Frame>>(evolved_fields).get(mu, nu) =
+                            0.5 * (u_plus.get(mu, nu) + u_minus.get(mu, nu)) +
+                            get(gamma_2) * u_psi.get(mu, nu);
+      for (size_t i = 0; i < Dim; ++i) {
+        get<Tags::Phi<Dim, Frame>>(evolved_fields).get(i,mu,nu) =
+                            0.5 * (u_plus.get(mu, nu) - u_minus.get(mu, nu)) *
+                            unit_normal_one_form.get(i) + u_zero.get(i, mu, nu);
+      }
+    }
+  }
+
+  return evolved_fields;
+}
+
+
+namespace {
+template <size_t Dim, typename Frame, typename DataType>
+struct CharacteristicFieldsComputeBuffer;
+
+template <size_t Dim, typename Frame>
+struct CharacteristicFieldsComputeBuffer<Dim, Frame, DataVector> {
+  private:
+    // We make one giant allocation so that we don't thrash the heap.
+    Variables<tmpl::list<::Tags::TempI<0, Dim, Frame>,
+                         ::Tags::Tempaa<1, Dim, Frame>,
+                         ::Tags::Tempiaa<2, Dim, Frame>>> buffer_;
+
+  public:
+    explicit CharacteristicFieldsComputeBuffer(const size_t size) noexcept :
+      buffer_(size, 0.), normal(get<::Tags::TempI<0, Dim, Frame>>(buffer_)),
+      phi_dot_normal(get<::Tags::Tempaa<1, Dim, Frame>>(buffer_)),
+      phi_dot_projection_tensor(get<::Tags::Tempiaa<2, Dim, Frame>>(buffer_)) {}
+
+    tnsr::I<DataVector, Dim, Frame>& normal;
+    tnsr::aa<DataVector, Dim, Frame>& phi_dot_normal;
+    tnsr::iaa<DataVector, Dim, Frame>& phi_dot_projection_tensor;
+};
+}  // namespace
+
+template <size_t Dim, typename Frame>
+typename Tags::CharacteristicFields<Dim, Frame>::type
+CharacteristicFieldsCompute<Dim, Frame>::function(
+    const Scalar<DataVector>& gamma_2,
+    const tnsr::aa<DataVector, Dim, Frame>& spacetime_metric,
+    const tnsr::II<DataVector, Dim, Frame>& inverse_spatial_metric,
+    const tnsr::aa<DataVector, Dim, Frame>& pi,
+    const tnsr::iaa<DataVector, Dim, Frame>& phi,
+    const tnsr::i<DataVector, Dim, Frame>& unit_normal_one_form) noexcept {
+  const size_t n_pts = unit_normal_one_form.begin()->size();
+  // Use a Variables to reduce total number of allocations. This is especially
+  // important in a multithreaded environment.
+  CharacteristicFieldsComputeBuffer<Dim, Frame, DataVector> buffer(n_pts);
+
+  // normal vector: n^i = g^{ij} n_j
+  buffer.normal =
+      raise_or_lower_index(unit_normal_one_form, inverse_spatial_metric);
```

Comment:
Why not pass in unit_normal_vector as well? Then someone else raises it once, and your'e done, instead of potentially raising it every time this function is called.

Reply:
good point, especially since this is a function which can be called multiple times in the same cell at the same time step. changed